File: src/main/java/nz/co/searchwellington/filters/attributesetters/PublisherPageAttributeSetter.java
package nz.co.searchwellington.filters.attributesetters;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.servlet.http.HttpServletRequest;

import nz.co.searchwellington.model.Website;
import nz.co.searchwellington.repositories.HibernateResourceDAO;

import org.apache.log4j.Logger;

public class PublisherPageAttributeSetter implements AttributeSetter {
	
	private static Logger log = Logger.getLogger(PublisherPageAttributeSetter.class);
	
	private static Pattern publisherPagePathPattern = Pattern.compile("^/(.*?)(/(comment|geotagged))?(/(edit|save|rss|json))?$");
	
	private HibernateResourceDAO resourceDAO;
	
	public PublisherPageAttributeSetter(HibernateResourceDAO resourceDAO) {
		this.resourceDAO = resourceDAO;
	}
	
	@Override
	public boolean setAttributes(HttpServletRequest request) {
		log.debug("Looking for single publisher path");
		Matcher contentMatcher = publisherPagePathPattern.matcher(request.getPathInfo());
		if (contentMatcher.matches()) {
			final String match = contentMatcher.group(1);
			log.debug("'" + match + "' matches content");
			log.debug("Looking for publisher '" + match + "'");
			Website publisher = (Website) resourceDAO.getPublisherByUrlWords(match);
			if (publisher != null) {
				log.info("Setting publisher: " + publisher.getName());
				request.setAttribute("publisher", publisher);
				request.setAttribute("resource", publisher);
				return true;			
			}
		}
		return false;
	}
	
}

File: src/main/scala/nz/co/searchwellington/filters/attributesetters/PublisherPageAttributeSetter.scala
package nz.co.searchwellington.filters.attributesetters

import jakarta.servlet.http.HttpServletRequest
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.filters.RequestPath
import nz.co.searchwellington.filters.attributesetters.PublisherPageAttributeSetter.PUBLISHER_ATTRIBUTE
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import java.util.regex.Pattern
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

@Component
class PublisherPageAttributeSetter @Autowired()(mongoRepository: MongoRepository)
  extends AttributeSetter with ReasonableWaits {

  private val publisherPagePathPattern = Pattern.compile("^/(.*?)(/(geotagged|.*?-.*?))?(/(rss|json))?$")

  def setAttributes(request: HttpServletRequest): Future[Map[String, Any]] = {
    val contentMatcher = publisherPagePathPattern.matcher(RequestPath.getPathFrom(request))
    if (contentMatcher.matches) {
      val publisherUrlWords = contentMatcher.group(1)
      if (publisherUrlWords.trim.nonEmpty && !publisherUrlWords.contains("+")) {
        mongoRepository.getWebsiteByUrlwords(publisherUrlWords).map { maybeWebsite =>
          maybeWebsite.map { publisher =>
            Map(
              PUBLISHER_ATTRIBUTE -> publisher
            )
          }.getOrElse {
            Map.empty
          }
        }
      } else {
        Future.successful(Map.empty)
      }
    } else {
      Future.successful(Map.empty)
    }
  }

}

object PublisherPageAttributeSetter {
  val PUBLISHER_ATTRIBUTE = "publisher"
}

